### DISCUSSION TITLE
Autodraw flowsheets?

### DISCUSSION THREAD

**Discussion by michox** [oemof/tespy#413] [Ideas]
Hi, I have seen that there are svg files referenced to each component. I was wondering if it would be possible to automatically connect these svgs to draw the full flowsheet of a network? It could be annotated with the labels and the attributes and that surely would help a lot with debugging!

**Comment by fwitte** [comment]
Hi Michael,

thank you for reaching out, that sounds like an excellent idea.

On my side I never really had the time to implement something like this. I was thinking something based on or related to [draw.io](https://draw.io/), which seems to have a lot of useful features. Basically generating the TESPy code from a draw.io drawing seems more straight forward to me: You'd have to make a connection between different shapes and the TESPy component objects.

But the other way round I am not as sure. There should be methods for that I'd guess, but I do not really know at the moment. Are you thinking to work on something like this?

Best

Francesco

**Reply by michox** [reply]
Draw.io is a good idea, it is quite versatile and can be implemented in ipynb books and the users can fiddle with the layout afterwards, since that is probably going to be the most difficult part. To be honest, I don't think I am a good enough python programmer to figure this out, but I did a bit of digging about python packages one could use, and I've asked GPT4 to come up with a first draft of a function that generates a drawio xml file. After a bit of back and forth with GPT4 it came up with this code, which looks like it's going in the right direction, but I won't have the time to test or implement this properly... This would require each component to have an attribute with the svg path. Note that this will not handle proper connections, but rather just randomly connect the links to the border of the svg. For proper handling, one would need to manually create a drawio library or find a suitable one. However, you will probably want to push around the nodes anyways, as this will just arrange everything in a circle.
```
import pandas as pd
import base64
from N2G import drawio_diagram
from tespy.networks import network

def generate_drawio_diagram(tespy_network: network, filename: str):
    # Create a new draw.io diagram
    diagram = drawio_diagram()
    diagram.add_diagram("TESPy Network")

    # Function to load and encode SVG file in Base64
    def encode_svg(svg_path: str) -> str:
        with open(svg_path, "rb") as f:
            svg_data = f.read()
        return base64.b64encode(svg_data).decode("utf-8")

    # Iterate through the components and connections
    for _, component in tespy_network.comps.iterrows():
        label = component.label
        # Set the attributes for the component
        for attr, value in component.attributes.items():
            label += f"\n{attr}: {value:.2f}"
        # Add the component node to the diagram
        diagram.add_node(
            id=component.name, 
            label=label, 
            style=f"image=data:image/svg+xml,{encode_svg(component.svg_path)}"
            )

    for _, connection in tespy_network.conns.iterrows():

        label = connection.label
        # Set the attributes for the connection
        for attr, value in connection.attributes.items():
            label += f"\n{attr}: {value:.2f}"

        # Add the link between the source and target components
        diagram.add_link(connection.source.name, connection.target.name, label=label, src_label=connection.source_id, trgt_label=connection.target_id)
            
    # adjust the layout of the diagram
    diagram.layout(algo="circular")

    # if filename does not have the xml or drawio extension, add it
    if not filename.endswith(".xml") and not filename.endswith(".drawio"):
        filename += ".drawio"

    # Save the diagram as an XML file
    diagram.dump_file(filename=filename, folder="./")
```

**Reply by michox** [reply]
and here is a python script (also AI generated) that should add a static method that the generate_drawio_diagram function could call to get the filepath to all components files, although it could be easier to simply use the existing "component()" method and generate the that way. I am also not sure, how to handle the relative position situation.
```python 
import os

# Define the folder containing the .py files
folder = "path/to/folder"

# Iterate through the .py files in the folder
for filename in os.listdir(folder):
    if filename.endswith(".py"):
        with open(os.path.join(folder, filename), "r") as file:
            content = file.readlines()

        # Find the first occurrence of the image string
        image_string = None
        for i, line in enumerate(content):
            if ".. image:: /api/_images/" in line:
                image_string = line.strip().split(".. image:: /api/_images/")[1].split(".svg")[0]
                break

        if image_string:
            # Find the first occurrence of the "@staticmethod" string
            staticmethod_index = None
            for i, line in enumerate(content):
                if "@staticmethod" in line:
                    staticmethod_index = i
                    break

            if staticmethod_index is not None:
                # Insert the new lines
                new_lines = [
                    "    @staticmethod\n",
                    "    def svg_path():\n",
                    f"        return '/api/_images/{image_string}.svg'\n",
                    "\n",
                ]
                for index, line in enumerate(new_lines):
                    content.insert(staticmethod_index + index, line)

                # Write the updated content back to the file
                with open(os.path.join(folder, filename), "w") as file:
                    file.writelines(content)
```

**Reply by fwitte** [reply]
Thank you for your suggestions, I will try to get a start from this on the weekend :).

Best

**Reply by michox** [reply]
Hi @fwitte, are you still interested in implementing this idea?

**Reply by fwitte** [reply]
Hi, not myself really, but maybe we could gather forces for this topic... If somebody would take the lead I'd be happy to join! E.g. see https://github.com/oemof/tespy/discussions/488, I really like that draw.io template. Since that produces json, one could build a tespy network based on that. This would be the other way around, but still a good starting point (arguably also the better one, because drawing the process and then running it sounds more convenient to me than coding it and let the flowsheet be drawn from that for most of the users).
